node/src/tasks/bsp_upload_file.rs

```diff
@@ -260,69 +242,21 @@ where
                 .map_err(|_| anyhow!("Failed to register peer file."))?;
         }
 
-        // Wait for the transaction to be included in a block.
-        let mut block_hash = None;
-        // TODO: Consider adding a timeout.
-        while let Some(tx_result) = tx_watcher.recv().await {
-            // Parse the JSONRPC string, now that we know it is not an error.
-            let json: serde_json::Value = serde_json::from_str(&tx_result).map_err(|_| {
-                anyhow!("The result, if not an error, can only be a JSONRPC string; qed")
-            })?;
-
-            debug!(target: LOG_TARGET, "Transaction information: {:?}", json);
-
-            // Checking if the transaction is included in a block.
-            // TODO: Consider if we might want to wait for "finalized".
-            // TODO: Handle other lifetime extrinsic edge cases. See https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/transaction-pool/api/src/lib.rs#L131
-            if let Some(in_block) = json["params"]["result"]["inBlock"].as_str() {
-                block_hash = Some(H256::from_str(in_block)?);
-                let subscription_id = json["params"]["subscription"]
-                    .as_number()
-                    .ok_or_else(|| anyhow!("Subscription should exist and be a number; qed"))?;
-
-                // Unwatch extrinsic to release tx_watcher.
-                self.storage_hub_handler
-                    .blockchain
-                    .unwatch_extrinsic(subscription_id.to_owned())
-                    .await?;
-
-                // Breaking while loop.
-                // Even though we unwatch the transaction, and the loop should break, we still break manually
-                // in case we continue to receive updates. This should not happen, but it is a safety measure,
-                // and we already have what we need.
-                break;
-            }
-        }
+        // Build extrinsic.
+        let call =
+            storage_hub_runtime::RuntimeCall::FileSystem(pallet_file_system::Call::bsp_volunteer {
+                location: event.location.clone(),
+                fingerprint: fingerprint.into(),
+            });
 
-        // Get the extrinsic from the block, with its events.
-        let block_hash = block_hash.ok_or_else(
-            || anyhow!("Block hash should exist after waiting for extrinsic to be included in a block; qed")
-        )?;
-        let extrinsic_in_block = self
-            .storage_hub_handler
+        self.storage_hub_handler
             .blockchain
-            .get_extrinsic_from_block(block_hash, tx_hash)
+            .send_extrinsic(call)
+            .await?
+            .with_timeout(Duration::from_secs(60))
+            .watch_for_success(&self.storage_hub_handler.blockchain)
```

Comment:
Why do you have to pass in blockchain handle when it is already accessible from `self` within the `watch_for_success`?

Reply:
I didn't want to pollute `SubmittedTransaction`.